File: ICPC/Trial-JP/2020/m.cpp
#include <iostream>
#include <queue>

using namespace std;

int pow3[15];
int dp[14348907];

int check(int a, int k) {
    a = a / pow3[k];
    return a % 3;
}

int main(){
    int q, m, l, i, j, k, mask;
    string s;
    queue<int> que;
    cin >> q >> m;
    for(i = 0;i < 15;i++){
        pow3[i] = pow(3, i);
    }
    for(i = 0;i < q;i++){
        cin >> s;
        for(j = 0;j < m; j++){
            l += stoi(s[j]) * pow3[j];
        }
        cout << dp[l];
        if(dp[l] == 0){
            que.push(l);
            while(!que.empty()){
                mask = que.front(); que.pop();
                for(k = 0; k < m; k++){
                    if(check(mask, k)){
                        if(dp[mask-pow3[k]]) continue;
                        dp[mask-pow3[k]] = 1;
                        que.push(mask-pow3[k]);
                    }
                }
            }
        }
    }
}

File: ICPC/Trial-JP/2020/m.py
def check(a,k):
    a = a // pow3[k]
    return a % 3

from collections import deque

q,m = map(int, input().split())
dp = [0] * (3 ** 15)
pow3 = [pow(3,i) for i in range(15)]
ans = [None] * q
for i in range(q)q  a:
    s = input()
    l = [int(i) - 1 for i in s]
    tmp = sum((l[i] * pow3[i] for i in range(m)))
    ans[i] = str(dp[tmp])
    if dp[tmp] == 0:
        q = deque([tmp])
        while q:
            mask = q.popleft()
            for k in range(m):
                if check(mask, k):
                    x = mask-pow3[k]
                    if dp[x]: continue
                    dp[x] = 1
                    q.append((x))
print("".join(ans))